### DISCUSSION TITLE
Correct use of HardwareSerial

### DISCUSSION THREAD

**Discussion by brightproject** [stm32duino/Arduino_Core_STM32#2226] [Libraries] [answered]
Good day @fpistm
I use this wonderful [framework](https://github.com/stm32duino/Arduino_Core_STM32/) to build firmware for `STM32F401/411`.
For my application, I need to use the `uBlox` receiver `GPS`, which works over `UART`.
I allocated `Serial` for debugging, and `UART2` for communication with `MCU` and`GPS` module.
The problem is using a third-party [library](https://github.com/bolderflight/ublox) to work with the `UBX` protocol.
My usual code.
Previously, this code always worked and I could use two or more serial ports for both input and output.
```
#define SERIAL_BAUD 115200
#define GNSS_BAUD 115200

HardwareSerial Serial2(USART2); // PA2(TX2), PA3(RX2)

void setup() 
{
  Serial.begin(SERIAL_BAUD);
  Serial2.begin(GNSS_BAUD);
}

void loop() {
Serial.println("Serial 1");
Serial2.println("Serial 2");

// if (Serial2.available() > 0)
// Serial.write(Serial2.read());
// Serial2.write(Serial.read());

}
```

My code for `uBlox GPS`:
```
#include <ubx.h>

#define SERIAL_BAUD 115200
#define GNSS_BAUD 115200

// #define USART2 Serial2

// extern HardwareSerial USART2;
HardwareSerial Serial2(USART2); // PA2(TX2), PA3(RX2)
// HardwareSerial Serial2(PA2, PA3);
// HardwareSerial& GPS = Serial2;

bfs::Ubx gnss(&Serial2);
//bfs::Ubx gnss(&GPS);

void setup() {
  /* Serial to display data */
  Serial.begin(SERIAL_BAUD);
  while(!Serial) {}
  gnss.Begin(GNSS_BAUD);
}

void loop() {

  delay(20);

  if(gnss.Read()) {
    Serial.print(gnss.fix());
    Serial.print("\t");
    Serial.print(gnss.num_sv());
    Serial.print("\t");
    Serial.print(gnss.lat_deg(), 6);
    Serial.print("\t");
    Serial.print(gnss.lon_deg(), 6);
    Serial.print("\t");
    Serial.print(gnss.alt_wgs84_m(), 2);
    Serial.print("\n");
  }
}
```
The code compiles without errors, but nothing is output to the serial port, I have some suspicions.
![compileOK](https://github.com/stm32duino/Arduino_Core_STM32/assets/1788098/5d2bec4d-1468-4da1-a857-a97191ec28fa)
First, in the sketch I declare the `HardwareSerial` and pass only a pointer to the port serial to the `UBX` library.
But the library itself also creates an instance for working with the `HardwareSerial` and it seems to me that this is the reason for the `serial ports` not working.
https://github.com/bolderflight/ublox/blob/b472d38bca21cc37843a04b9422692edd5f5d6e2/src/ubx.h#L52
`explicit Ubx(HardwareSerial* bus) : bus_(bus) {}`
The difficulty is that the [authors](https://github.com/bolderflight/ublox/issues/21) wrote the [library](https://github.com/bolderflight/ublox) mainly for the `Teensy` microcontroller.
And everything works on it without problems.
No matter how hard I tried to remove the `HardwareSerial` port declaration, I always got an error.
How can I correctly declare a `HardwareSerial` port so as not to rewrite the library?
I hope my explanations will not confuse you, because I myself am already quite confused.

**Comment by fpistm** [comment]
Hi @brightproject 
As a first feedback, seems you used the same U(S)ART. By default Serial is mapped to USART2 for STM32F401:
https://github.com/stm32duino/Arduino_Core_STM32/blob/63d30964d41809b3f704f0cab8c5b64870fdcc0d/variants/STM32F4xx/F401R(B-C-D-E)T/variant_generic.h#L147-L152

So I guess here it is the first mistake.
For me the library does not instantiate HardwareSerial, you provide only a pointer to it.

**Comment by brightproject** [comment]
Thank you for your answer, but it seems to me that you did not quite understand me, and you did not ask any questions🙂
Below I have simplified the code to make it more clear.
By default, in `STM32` you can only use serial port 1 and we designate its `Serial`, if we want to use `serial 2`, then we just don’t declare it.
You must specify the use of `HardwareSerial`.
If you do not specify this explicitly, then calling the code
`Serial2.Begin(SERIAL2_BAUD )`;
in the setup section will lead to a compilation error, which is logical.
```
#define SERIAL_BAUD 115200
#define SERIAL2_BAUD 115200

HardwareSerial Serial2(USART2); 

void setup() {
  Serial.begin(SERIAL_BAUD);
  Serial2.Begin(SERIAL2_BAUD );
}

void loop() {
// somethings
}
```
Moving on, I have a third-party library, which inside itself has an algorithm for determining the `HardwareSerial`.
The library input must be supplied with the name of the port serial to be used.
But, according to the description above, I cannot submit just `Serial2` as input to the library, because I need to declare it first through
`HardwareSerial` Serial2(USART2); 
And it turns out that I declare the `HardwareSerial` twice - which is generally not true and possible, which is why I have problems with the library.
```
void Ubx::Config(HardwareSerial* bus) {
  bus_ = bus;
}
```
Specify this code in the library
`USART2.Begin(SERIAL2_BAUD` )
This is also not possible, there will be a compilation error.

**Reply by fpistm** [reply, marked as answer]
> By default, in `STM32` you can only use serial port 1 and we designate its `Serial`

No. `Serial` is simply an alias to the specified `Serialx` defined in the variant. 
Ex for generic F401RB: 
https://github.com/stm32duino/Arduino_Core_STM32/blob/c83db2e4f6b4b8d3bbedce67d2156a58023bc432/variants/STM32F4xx/F401R(B-C-D-E)T/variant_generic.h#L140-L152
So `Serial` is `Serial1`.

But for Generic F412ZG:
https://github.com/stm32duino/Arduino_Core_STM32/blob/c83db2e4f6b4b8d3bbedce67d2156a58023bc432/variants/STM32F4xx/F412Z(E-G)(J-T)/variant_generic.h#L228-L240
`Serial` is `Serial2`

I advise you read the wiki:
https://github.com/stm32duino/Arduino_Core_STM32/wiki/API#hardwareserial

As stated your first code uses the same `Serial` instance. Which target you select? Generic STM32F401RB ? other?

> And it turns out that I declare the `HardwareSerial` twice - which is generally not true and possible, which is why I have problems with the library.
> 
> ```
> void Ubx::Config(HardwareSerial* bus) {
>   bus_ = bus;
> }
> ```

No it is a pointer the third library do not "declare `HardwareSerial` twice". Several library worked like this to ease usage. Maybe a reference would be better but it is an other discussion.
It is well documented in the library you used:
https://github.com/bolderflight/ublox?tab=readme-ov-file#methods

![image](https://github.com/stm32duino/Arduino_Core_STM32/assets/20641798/a78d9e4d-ff2f-464c-8495-ce44f14b5d43)


> #define SERIAL_BAUD 115200
> #define SERIAL2_BAUD 115200
> 
> HardwareSerial Serial2(USART2); 
> 
> void setup() {
>   Serial.begin(SERIAL_BAUD);
>   Serial2.Begin(SERIAL2_BAUD );
> }
> 
> void loop() {
> // somethings
> }

Here this code is not correct for generic STM32F401RB as it is not `Begin` but `begin`.
Then after correct this the error is clear:
`core/core.a(HardwareSerial.cpp.o):(.bss.Serial2+0x0): multiple definition of `Serial2'; `
which is normal, as stated `Serial` is `Serial2`.

> Specify this code in the library
> `USART2.Begin(SERIAL2_BAUD` )
> This is also not possible, there will be a compilation error.

This is normal `USART2` name is an CMSIS device definition:
https://github.com/stm32duino/Arduino_Core_STM32/blob/c83db2e4f6b4b8d3bbedce67d2156a58023bc432/system/Drivers/CMSIS/Device/ST/STM32F4xx/Include/stm32f401xe.h#L759C1-L759C1

it seems you don't understand what you do.
So if you use the generic STM32F401RE (same for Nucleo), `Serial` is `Serial2` so no need to redefine it.
If you used `Serial` for "debug", nothing to do.
Then you have to declare and use a new instance for your GPS, ex `USART1` using `PA9`/`PA10`.
`HardwareSerial Serial1(USART1);`
or 
`HardwareSerial Serial1(PA10, PA9);`
or you can also use an custom name:
`HardwareSerial myGPS(PA10, PA9);`

So, a correct code would be;
```C++
#include "ubx.h"

#define SERIAL_BAUD 115200
#define GNSS_BAUD 921600

HardwareSerial myGNSS(PA10, PA9);

/* Ublox object, GNSS on USART1 pins PA9/PA10 name myGNSS */
bfs::Ubx gnss(&myGNSS);

void setup() {
  Serial.begin(SERIAL_BAUD);
  // Ublox instance will call begin of the myGNSS instance.
  // myGNSS.begin(GNSS_BAUD);
  gnss.Begin(GNSS_BAUD);
}

void loop() {
  // somethings
}
```
